### ISSUE TITLE
Question: using setState to access latest state in event handler

### ISSUE THREAD

**Issue by 832bb9** [facebook/react#20111]
Is it bad to use setState for getting current state in async event handlers?

![25C4002A-8A34-49CA-A3EE-898419EF98DF](https://user-images.githubusercontent.com/31843788/97357104-eecac700-18a1-11eb-9af9-ddcd04d9c2be.jpeg)

 It works good but feels wrong, so you might say why it is wrong. I know recommended approach is to use refs, but why have redundant work with sync ref and state. If it is safe to access current state due function updater for setState so you can omit unnecessary state deps for callback, why it might be not safe to get current state in async handler?

Example:
You have “follow” user button. When you click it update local state optimistically and make server request. While server request pending you might click again and change local state, but prevent making new requests. When request is resolved you need to compare latest local state and expected server state and repeat request if they aren’t equal.

**Comment by kachkaev** [comment]
For a bug report, please use the issue template: https://github.com/facebook/react/blob/master/.github/ISSUE_TEMPLATE/bug_report.md

For questions and help, please follow https://reactjs.org/community/support.html

**Comment by 832bb9** [comment]
Yeah, I checked this doc page before creating issue, but @gaearon specifically said to create issue if there is question only maintainers can answer (https://twitter.com/dan_abramov/status/1321061865829576705).

**Comment by a-c-sreedhar-reddy** [comment]
But should not function to setState be pure?

**Comment by 832bb9** [comment]
That is the question. Should it?

**Comment by 832bb9** [comment]
Promise will resolve only once. The question if it will be resolved with expected value.

**Comment by a-c-sreedhar-reddy** [comment]
yeah promise resolve would not have an effect if run second time.

 I guess it would resolve with the latest state as expected.

**Comment by jordyvandomselaar** [comment]
Hi,

Wouldn't you be able to access `state` directly? Components have `this.state` and  `useState` always returns an array where the first item is the actual state and the second item is the setter.

**Comment by jordyvandomselaar** [comment]
Could you make a small codesandbox with what you're trying to do? That makes it easier to see what you mean and propose a fix =)

I'm almost certain `useEffect` or `componentDidUpdate` might solve your usecase but I'm not sure.

**Comment by lstkz** [comment]
Here is a similar discussion https://github.com/facebook/react/issues/15337
You can create your own version of `useState` (which is simple to implement):

```ts
const [state, setState, getState] = useStateExt(null);
```

It would be good to have it in the standard API because there are valid use cases for it.
For example: when you use `setInterval` you can't have access the latest state. Invalidating the callback can be problematic.

**Comment by 832bb9** [comment]
Yeah, this is similar to what I am doing now, but instead of extending `useState` I am using `useStableCallback` hook which implements class-like-components semantics (have stable identity but returns latest render scope). The question is if it is valid to use current react API for this case without any additional work.

**Comment by 832bb9** [comment]
> Could you make a small codesandbox with what you're trying to do? That makes it easier to see what you mean and propose a fix =)
> 
> I'm almost certain `useEffect` or `componentDidUpdate` might solve your usecase but I'm not sure.

Please check description. I need accept current state in async handler in a while after calling it. Handlers scope has access only to render scope before calling, so if state would be changed until I check my handler can’t accept latest state. It can be resolved via refs (this is what I am doing now), but if I can accept it without working with them it would be better.

**Comment by jordyvandomselaar** [comment]
@832bb9 Is it possible to do the async work first and then after that access the state instead of sending state to the async method, doing some work and then accessing the state?

**Comment by 832bb9** [comment]
Still the question: should state updater be pure function? For example here https://github.com/mantinedev/mantine/blob/master/src/mantine-hooks/src/use-local-storage-value/use-local-storage-value.ts. We can’t set state to local storage before we can calculate it based on previous value, but can we do something in updater or we should use effect after component rerender?

**Comment by eps1lon** [comment]
> Still the question: should state updater be pure function?

Yes. [StrictMode](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects) already tests this by invoking updater functions twice.

@832bb9  Could you create a codesandbox illustrating the concrete use case for a getter function? I think your feature request is already covered by https://github.com/facebook/react/issues/14543. If not, could you clarify what the difference to https://github.com/facebook/react/issues/14543 is?
